core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java

```diff
@@ -122,22 +123,29 @@ public static Object parseStoredFieldsValue(XContentParser parser) throws IOExce
      * @param delimiter   the delimiter to use to splits the field's name
      * @param objectClass the object class of the object to parse
      * @param <T>         the type of the object to parse
+     * @param lenient     set to <tt>true</tt> if the method should throw a ParsingException in no delimeter was found
      * @return the parsed object
      * @throws IOException if anything went wrong during parsing or if the type or name cannot be derived
      *                     from the field's name
      */
-    public static <T> T parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass) throws IOException {
+    public static <T> Optional<T> parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass, boolean lenient)
+            throws IOException {
         String currentFieldName = parser.currentName();
         if (Strings.hasLength(currentFieldName)) {
             int position = currentFieldName.indexOf(delimiter);
             if (position > 0) {
                 String type = currentFieldName.substring(0, position);
                 String name = currentFieldName.substring(position + 1);
-                return parser.namedObject(objectClass, type, name);
+                return Optional.of(parser.namedObject(objectClass, type, name));
             }
         }
-        throw new ParsingException(parser.getTokenLocation(), "Cannot parse object of class [" + objectClass.getSimpleName()
-                + "] without type information. Set [" + RestSearchAction.TYPED_KEYS_PARAM + "] parameter on the request to ensure the"
-                + " type information is added to the response output");
+        if (lenient) {
+            parser.skipChildren();
```

Comment:
Fields and Arrays are already inserted randomly in all xContent objects (except for the exceptions defined in the test) in this PR, so we should already be covered.